components/renku_data_services/connected_services/blueprints.py

```diff
@@ -0,0 +1,213 @@
+        """Delete a specific OAuth2 Client."""
+
+        @authenticate(self.authenticator)
+        @only_admins
+        async def _delete(_: Request, provider_id: str, user: base_models.APIUser):
+            await self.connected_services_repo.delete_oauth2_client(user=user, provider_id=provider_id)
+            return HTTPResponse(status=204)
+
+        return "/oauth2/admin/providers/<provider_id>", ["DELETE"], _delete
+
+
+@dataclass(kw_only=True)
+class OAuth2ConnectionsBP(CustomBlueprint):
+    """Handlers for using OAuth2 connections."""
+
+    connected_services_repo: ConnectedServicesRepository
+    authenticator: base_models.Authenticator
+
+    def get_all(self) -> BlueprintFactoryResponse:
+        """List all OAuth2 connections."""
+
+        @authenticate(self.authenticator)
+        async def _get_all(_: Request, user: base_models.APIUser):
+            connections = await self.connected_services_repo.get_oauth2_connections(user=user)
+            return json(
+                [apispec.Connection.model_validate(c).model_dump(exclude_none=True, mode="json") for c in connections]
+            )
+
+        return "/oauth2/connections", ["GET"], _get_all
+
+    def get_one(self) -> BlueprintFactoryResponse:
+        """Get a specific OAuth2 connection."""
+
+        @authenticate(self.authenticator)
+        async def _get_one(_: Request, connection_id: str, user: base_models.APIUser):
+            connection = await self.connected_services_repo.get_oauth2_connection(
+                connection_id=connection_id, user=user
+            )
+            return json(apispec.Connection.model_validate(connection).model_dump(exclude_none=True, mode="json"))
+
+        return "/oauth2/connections/<connection_id>", ["GET"], _get_one
+
+    def get_account(self) -> BlueprintFactoryResponse:
+        """Get the account information for a specific OAuth2 connection."""
+
+        @authenticate(self.authenticator)
+        async def _get_account(_: Request, connection_id: str, user: base_models.APIUser):
+            account = await self.connected_services_repo.get_oauth2_connected_account(
+                connection_id=connection_id, user=user
+            )
+            return json(apispec.ConnectedAccount.model_validate(account).model_dump(exclude_none=True, mode="json"))
+
+        return "/oauth2/connections/<connection_id>/account", ["GET"], _get_account
+
+    def get_token(self) -> BlueprintFactoryResponse:
+        """Get the access token for a specific OAuth2 connection."""
+
+        @authenticate(self.authenticator)
+        async def _get_token(_: Request, connection_id: str, user: base_models.APIUser):
+            token = await self.connected_services_repo.get_oauth2_connection_token(
+                connection_id=connection_id, user=user
+            )
+            return json(token.dump_for_api())
+
+        return "/oauth2/connections/<connection_id>/token", ["GET"], _get_token
```

Comment:
This is intentional, I do not want to expose this endpoint. Ideally we would make it so that only sessions can generate tokens from this endpoint.

Reply:
there's not really a great way to enforce things like that, I had a similar problem in secrets storage and solved it by [denying the path in the gateway](https://github.com/SwissDataScienceCenter/renku-gateway/pull/703) so it's not accessible from outside the cluster.
I don't think we can have a better solution, there's the possibility of ingress rules based on path but since the ingress might be different in third party deployments, having it in the gateway is more consistent.
I do have the endpoint in the api spec in my case, but with a comment that it is cluster internal only

Reply:
I don't know if this is something we can do for sessions; they use the external https URL to do get the tokens, as they are not allowed to hit internal endpoints